opsdroid/database/matrix/__init__.py

```diff
@@ -0,0 +1,285 @@
+"""Database that uses the matrix connector."""
+
+import logging
+from contextlib import contextmanager
+
+from nio import RoomGetStateError, RoomGetStateEventError, RoomGetEventError
+from opsdroid.database import Database
+from opsdroid.connector.matrix.events import MatrixStateEvent, GenericMatrixRoomEvent
+from voluptuous import Any
+
+_LOGGER = logging.getLogger(__name__)
+CONFIG_SCHEMA = {
+    "default_room": str,
+    "single_state_key": Any(bool, str),
+    "should_encrypt": bool,
+}
+
+
+class DatabaseMatrix(Database):
+    """A module for opsdroid to allow memory to persist in matrix room state."""
+
+    def __init__(self, config, opsdroid=None):
+        """Start the database connection."""
+        super().__init__(config, opsdroid=opsdroid)
+        self.name = "matrix"
+        self.room = config.get("default_room", "main")
+        self._single_state_key = config.get("single_state_key", True)
+        self.should_encrypt = config.get("should_encrypt", True)
+        self._event_type = "dev.opsdroid.database"
+        self.should_migrate = True
+
+        _LOGGER.debug("Loaded matrix database connector.")
+
+    @property
+    def connector(self):
+        """Define the initiated matrix connector."""
+        return self.opsdroid.get_connector("matrix")
+
+    @property
+    def room_id(self):
+        """Define ID of the room to store data in."""
+        return self.connector.lookup_target(self.room)
+
+    @property
+    def is_room_encrypted(self):
+        """Check if room is encrypted."""
+        return (
+            self.connector._allow_encryption
+            and self.room_id in self.connector.connection.store.load_encrypted_rooms()
+        )
+
+    async def migrate_database(self):
+        """Migrate existing 'opsdroid.database' state event to 'dev.opsdroid.database' event."""
+
+        data = await self.connector.connection.room_get_state(room_id=self.room_id)
+        if isinstance(data, RoomGetStateError):
+            _LOGGER.error(
+                f"Error migrating from opsdroid.database to {self._event_type} in room {self.room_id}: {data.message}({data.status_code})"
+            )
+            return
+
+        if not any(event["type"] == self._event_type for event in data.events):
+            for event in data.events:
+                if event["type"] == "opsdroid.database" and event["content"]:
+                    await self.connector.connection.room_put_state(
+                        room_id=self.room_id,
+                        event_type=self._event_type,
+                        content=event["content"],
+                        state_key=event["state_key"],
+                    )
+                    await self.connector.connection.room_redact(
+                        room_id=self.room_id, event_id=event["event_id"],
+                    )
+
+        self.should_migrate = False
+
+    async def _verify_and_get(self, state_key, value_dict):
```

Comment:
I think this method could use a little readjustment, both to make it's error handling more applicable and to make it equally useful to `put` and `get`.

I think that restricting this to just getting an existing value out of the store and decrypting it if needed would be better than worrying about deduplicating with a new value, as that behaviour only belongs in `put` not `get`?

I would then call it something like `_get_and_decrypt`?